File: examples/apps/flutter/lib/core/network/api_client.dart
import 'dart:async';
import 'dart:convert';

import 'package:app_prueba_mobilelab/core/config/api_config.dart';
import 'package:app_prueba_mobilelab/core/errors/app_exception.dart';
import 'package:http/http.dart' as http;

class ApiClient {
  ApiClient({http.Client? client}) : _client = client ?? http.Client();

  final http.Client _client;
  String? token;

  Map<String, String> get _headers => {
    'Accept': 'application/json',
    'Content-Type': 'application/json',
    if (token != null) 'Authorization': 'Bearer $token',
  };

  Future<dynamic> get(String path) => _send('GET', path);

  Future<dynamic> post(String path, {Map<String, dynamic>? body}) =>
      _send('POST', path, body: body);

  Future<dynamic> put(String path, {Map<String, dynamic>? body}) =>
      _send('PUT', path, body: body);

  Future<dynamic> patch(String path, {Map<String, dynamic>? body}) =>
      _send('PATCH', path, body: body);

  Future<dynamic> delete(String path, {Map<String, dynamic>? body}) =>
      _send('DELETE', path, body: body);

  Future<dynamic> _send(
    String method,
    String path, {
    Map<String, dynamic>? body,
  }) async {
    final request = http.Request(method, Uri.parse('${ApiConfig.baseUrl}$path'))
      ..headers.addAll(_headers);
    if (body != null) request.body = jsonEncode(body);

    try {
      final streamed = await _client
          .send(request)
          .timeout(const Duration(seconds: 12));
      final response = await http.Response.fromStream(streamed);

      if (response.statusCode < 200 || response.statusCode >= 300) {
        throw AppException(
          'La API respondió con HTTP ${response.statusCode}.',
          statusCode: response.statusCode,
        );
      }

      if (response.bodyBytes.isEmpty) return null;
      return jsonDecode(utf8.decode(response.bodyBytes));
    } on TimeoutException {
      throw const AppException('La API tardó demasiado en responder.');
    } on http.ClientException catch (error) {
      throw AppException('No se pudo conectar con MobileLab: ${error.message}');
    } on FormatException {
      throw const AppException('La API devolvió una respuesta JSON inválida.');
    }
  }
}

File: examples/apps/react-native/src/core/network/ApiClient.ts
import { API_BASE_URL } from '../config/apiConfig';
import { AppError } from '../errors/AppError';

type Method = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';

export class ApiClient {
  private token?: string;

  setToken(token?: string) {
    this.token = token;
  }

  get<T>(path: string) {
    return this.request<T>('GET', path);
  }

  post<T>(path: string, body?: unknown) {
    return this.request<T>('POST', path, body);
  }

  put<T>(path: string, body?: unknown) {
    return this.request<T>('PUT', path, body);
  }

  patch<T>(path: string, body?: unknown) {
    return this.request<T>('PATCH', path, body);
  }

  delete<T>(path: string, body?: unknown) {
    return this.request<T>('DELETE', path, body);
  }

  private async request<T>(method: Method, path: string, body?: unknown) {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 12000);
    try {
      const response = await fetch(`${API_BASE_URL}${path}`, {
        method,
        signal: controller.signal,
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
          ...(this.token ? { Authorization: `Bearer ${this.token}` } : {}),
        },
        body: body === undefined ? undefined : JSON.stringify(body),
      });
      const text = await response.text();
      const data = text ? JSON.parse(text) : undefined;
      if (!response.ok) {
        throw new AppError(
          data?.message ?? `La API respondió con HTTP ${response.status}.`,
          response.status,
        );
      }
      return data as T;
    } catch (error) {
      if (error instanceof AppError) {
        throw error;
      }
      if (error instanceof Error && error.name === 'AbortError') {
        throw new AppError('La API tardó demasiado en responder.');
      }
      throw new AppError('No se pudo conectar con MobileLab.');
    } finally {
      clearTimeout(timeout);
    }
  }
}
